File: src/ac/parser/hostapply.go
/*
 * Tencent is pleased to support the open source community by making 蓝鲸 available.
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

package parser

import (
	"net/http"
	"regexp"

	"configcenter/src/ac/meta"
)

// HostApplyAuthConfigs TODO
var HostApplyAuthConfigs = []AuthConfig{
	{
		Name:           "CreateHostApplyRuleRegex",
		Description:    "添加主机属性自动应用规则",
		Regex:          regexp.MustCompile(`^/api/v3/create/host_apply_rule/bk_biz_id/([0-9]+)/?$`),
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    BizIDFromURLGetter,
		BizIndex:       5,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.Update,
	}, {
		Name:           "UpdateHostApplyRuleRegex",
		Description:    "更新主机属性自动应用规则",
		Regex:          regexp.MustCompile(`^/api/v3/update/host_apply_rule/([0-9]+)/bk_biz_id/([0-9]+)/?$`),
		HTTPMethod:     http.MethodPut,
		BizIDGetter:    BizIDFromURLGetter,
		BizIndex:       6,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.Update,
	}, {
		Name:           "DeleteModuleHostApplyRuleRegex",
		Description:    "删除模块场景下主机属性自动应用规则",
		Regex:          regexp.MustCompile(`^/api/v3/host/deletemany/module/host_apply_rule/bk_biz_id/([0-9]+)/?$`),
		HTTPMethod:     http.MethodDelete,
		BizIDGetter:    BizIDFromURLGetter,
		BizIndex:       7,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.Delete,
	}, {
		Name:           "GetHostApplyRuleRegex",
		Description:    "获取主机属性自动应用规则",
		Regex:          regexp.MustCompile(`^/api/v3/find/host_apply_rule/([0-9]+)/bk_biz_id/([0-9]+)/?$`),
		HTTPMethod:     http.MethodGet,
		BizIDGetter:    BizIDFromURLGetter,
		BizIndex:       6,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.DefaultHostApply,
	}, {
		Name:           "ListHostApplyRuleRegex",
		Description:    "列表查询主机属性自动应用规则",
		Regex:          regexp.MustCompile(`^/api/v3/findmany/host_apply_rule/bk_biz_id/([0-9]+)/?$`),
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    BizIDFromURLGetter,
		BizIndex:       5,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.DefaultHostApply,
	}, {
		Name:           "FindmanyModuleHostApplyTaskStatus",
		Description:    "查询模块场景下主机自动应用任务状态",
		Regex:          regexp.MustCompile(`^/api/v3/host/findmany/module/host_apply_plan/status`),
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    DefaultBizIDGetter,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.SkipAction,
	}, {
		Name:           "BatchUpdateOrCreateHostApplyRuleRegex",
		Description:    "批量创建/更新主机属性自动应用规则",
		Regex:          regexp.MustCompile(`^/api/v3/createmany/host_apply_rule/bk_biz_id/([0-9]+)/batch_create_or_update/?$`),
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    BizIDFromURLGetter,
		BizIndex:       5,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.Update,
	}, {
		Name:           "PreviewModuleApplyHostApplyRulePattern",
		Description:    "预览模块主机属性自动应用",
		Pattern:        "/api/v3/host/createmany/module/host_apply_plan/preview",
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    DefaultBizIDGetter,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.DefaultHostApply,
	}, {
		Name:           "RunHostApplyRuleOnModuleRegex",
		Description:    "模块场景下执行主机属性自动应用",
		Regex:          regexp.MustCompile(`^/api/v3/host/updatemany/module/host_apply_plan/run`),
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    DefaultBizIDGetter,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.Update,
	}, {
		Name:           "FindHostRelatedHostApplyRuleRegex",
		Description:    "查询主机关联的主机属性自动应用规则",
		Regex:          regexp.MustCompile(`^/api/v3/findmany/host_apply_rule/bk_biz_id/([0-9]+)/host_related_rules/?$`),
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    BizIDFromURLGetter,
		BizIndex:       5,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.DefaultHostApply,
	}, {
		Name:           "GetTemplateHostApplyStatusPattern",
		Description:    "获取模块所属服务模板是否开启了主机自动应用",
		Pattern:        "/api/v3/host/find/service_template/host_apply_status",
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    DefaultBizIDGetter,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.DefaultHostApply,
	}, {
		Name:           "GetServiceTemplateHostApplyRulePattern",
		Description:    "获取服务模板主机自动应用规则",
		Pattern:        "/api/v3/host/findmany/service_template/host_apply_rule",
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    DefaultBizIDGetter,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.DefaultHostApply,
	}, {
		Name:           "PreviewServiceTemplateApplyHostApplyRulePattern",
		Description:    "预览服务模版主机属性自动应用",
		Pattern:        "/api/v3/host/createmany/service_template/host_apply_plan/preview",
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    DefaultBizIDGetter,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.DefaultHostApply,
	}, {
		Name:           "GetModuleInvalidHostCountPattern",
		Description:    "获取模块下的主机自动应用失效主机数量",
		Pattern:        "/api/v3/host/findmany/module/host_apply_plan/invalid_host_count",
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    DefaultBizIDGetter,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.DefaultHostApply,
	}, {
		Name:           "GetServiceTemplateInvalidHostCountPattern",
		Description:    "获取服务模版下的主机自动应用失效主机数量",
		Pattern:        "/api/v3/host/findmany/service_template/host_apply_plan/invalid_host_count",
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    DefaultBizIDGetter,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.DefaultHostApply,
	}, {
		Name:           "GetServiceTemplateHostApplyRuleCountPattern",
		Description:    "获取服务模板配置的规则数",
		Pattern:        "/api/v3/host/findmany/service_template/host_apply_rule_count",
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    DefaultBizIDGetter,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.DefaultHostApply,
	}, {
		Name:           "GetModuleFinalRulesPattern",
		Description:    "获取模块的最终属性自动应用规则",
		Pattern:        "/api/v3/host/findmany/module/get_module_final_rules",
		HTTPMethod:     http.MethodPost,
		BizIDGetter:    DefaultBizIDGetter,
		ResourceType:   meta.HostApply,
		ResourceAction: meta.DefaultHostApply,
	},
}

// HostApply TODO
func (ps *parseStream) HostApply() *parseStream {
	return ParseStreamWithFramework(ps, HostApplyAuthConfigs)
}

File: ui/src/store/modules/api/host-apply.js
/*
 * Tencent is pleased to support the open source community by making 蓝鲸 available.
 * Copyright (C) 2017-2022 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable no-unused-vars */
import $http from '@/api'
import CombineRequest from '@/api/combine-request'

const state = {
  propertyConfig: {},
  ruleDraft: {},
  propertyList: []
}

const getters = {
  configPropertyList: (state) => {
    state.propertyList.forEach((property) => {
      // 兼容通用方法
      property.options = property.option
      // 自定义字段空间
      // eslint-disable-next-line no-underscore-dangle
      property.__extra__ = {
        visible: true
      }
    })
    const enabledList = state.propertyList.filter(item => item.host_apply_enabled)
    const disabledList = state.propertyList.filter(item => !item.host_apply_enabled)
    return [...enabledList, ...disabledList]
  }
}

const actions = {
  async getRules({ commit, state, dispatch }, { bizId, params, config }) {
    const reqs = await CombineRequest.setup(Symbol(), (ids) => {
      const subParams = { ...params, bk_module_ids: ids }
      return $http.post(`findmany/host_apply_rule/bk_biz_id/${bizId}`, subParams, config)
    }, { segment: 500 }).add(params.bk_module_ids)

    const results = []
    const res = await Promise.all(reqs)
    res.forEach(({ info: list = [] }) => {
      results.push(...list)
    })

    return {
      count: results.length,
      info: results
    }
  },
  async getTemplateRules({ commit, state, dispatch }, { bizId, params, config }) {
    const reqs = await CombineRequest.setup(Symbol(), (ids) => {
      const subParams = { ...params, service_template_ids: ids }
      return $http.post('host/findmany/service_template/host_apply_rule', subParams, config)
    }, { segment: 500 }).add(params.service_template_ids)

    const results = []
    const res = await Promise.all(reqs)
    res.forEach(({ info: list = [] }) => {
      results.push(...list)
    })

    return {
      count: results.length,
      info: results
    }
  },
  getApplyPreview({ commit, state, dispatch }, { params, config }) {
    return $http.post('host/createmany/module/host_apply_plan/preview', params, config)
  },
  getTemplateApplyPreview({ commit, state, dispatch }, { params, config }) {
    return $http.post('host/createmany/service_template/host_apply_plan/preview', params, config)
  },
  runApply({ commit, state, dispatch }, { params, config }) {
    return $http.post('host/updatemany/module/host_apply_plan/run', params, config)
  },
  runTemplateApply({ commit, state, dispatch }, { params, config }) {
    return $http.post('updatemany/proc/service_template/host_apply_plan/run', params, config)
  },
  getApplyTaskStatus({ commit, state, dispatch }, { params, config }) {
    return $http.post('host/findmany/module/host_apply_plan/status', params, config)
  },
  getTemplateApplyTaskStatus({ commit, state, dispatch }, { params, config }) {
    return $http.post('findmany/proc/service_template/host_apply_plan/status', params, config)
  },
  getTopopath({ commit, state, dispatch }, { bizId, params, config }) {
    return $http.post(`find/topopath/biz/${bizId}`, params, config)
  },
  setEnableStatus({ commit, state, dispatch }, { bizId, params, config }) {
    return $http.put(`module/host_apply_enable_status/bk_biz_id/${bizId}`, params, config)
  },
  setTemplateEnableStatus({ commit, state, dispatch }, { bizId, params, config }) {
    return $http.put(`updatemany/proc/service_template/host_apply_enable_status/biz/${bizId}`, params, config)
  },
  getConflictCount({ commit, state, dispatch }, { params, config }) {
    return $http.post('host/findmany/module/host_apply_plan/invalid_host_count', params, config)
  },
  getTemplateConflictCount({ commit, state, dispatch }, { params, config }) {
    return $http.post('host/findmany/service_template/host_apply_plan/invalid_host_count', params, config)
  },
  deleteRules({ commit, state, dispatch }, { bizId, params }) {
    return $http.delete(`host/deletemany/module/host_apply_rule/bk_biz_id/${bizId}`, params)
  },
  deleteTemplateRules({ commit, state, dispatch }, { bizId, params }) {
    return $http.delete(`deletemany/proc/service_template/host_apply_rule/biz/${bizId}`, params)
  },
  getProperties(context, { params, config }) {
    return $http.post('find/objectattr/host', params, config)
  },
  getHostRelatedRules(context, { bizId, params, config }) {
    return $http.post(`findmany/host_apply_rule/bk_biz_id/${bizId}/host_related_rules`, params, config)
  },
  searchNode(context, { bizId, params, config }) {
    return $http.post(`find/topoinst/bk_biz_id/${bizId}/host_apply_rule_related`, params, config)
  },
  getModuleApplyStatusByTemplate(context, { params, config }) {
    return $http.post('host/find/service_template/host_apply_status', params, config)
  },
  searchTemplateNode(context, { params, config }) {
    return $http.post('find/proc/service_template/host_apply_rule_related', params, config)
  },
  getTemplateRuleCount(context, { params, config }) {
    return $http.post('host/findmany/service_template/host_apply_rule_count', params, config)
  },
  getModuleFinalRules({ commit, state, dispatch }, { params, config }) {
    return $http.post('host/findmany/module/get_module_final_rules', params, config)
  }
}

const mutations = {
  setPropertyConfig(state, config) {
    state.propertyConfig = config
  },
  setRuleDraft(state, draft) {
    state.ruleDraft = draft
  },
  clearRuleDraft(state) {
    state.ruleDraft = {}
  },
  setPropertyList(state, list) {
    state.propertyList = list
  }
}

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations
}
